Feature/more floorplans

smits opened the pull request:
- Adds topics for all floorplan layers 
- Exposes all 3D reconstruction parameters.  

For the time being the occupancy grids are constructed from the raw floorplan data and the conversion to the ROS defaults for occupancy (0-100 instead of 0-255) has been disabled.

mcopejans commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_ros_node.h:
```diff
@@ -81,15 +84,25 @@ const std::string USE_FLOOR_PLAN_PARAM_NAME = "use_floor_plan";
+const std::string TANGO_3DR_MIN_NUM_VERTICES_PARAM_NAME = "reconstruction_min_num_vertices";
+const std::string TANGO_3DR_UPDATE_METHOD_PARAM_NAME = "reconstruction_update_method";
+const std::string TANGO_3DR_MAX_VOXEL_WEIGHT_PARAM_NAME = "reconstruction_max_voxel_weight";
+const std::string TANGO_3DR_FLOORPLAN_MAX_ERROR_PARAM_NAME = "reconstruction_floorplan_max_error";
```
Nit. rename param name values to `tango_3dr_floorplan_max_error` for consistency with param names above.

smits replied:
this parameter is really about the floorplan_max_error and only has an effect on the floorplan while the other have an effect on all 3D reconstruction feature. cfr https://developers.google.com/tango/apis/c/reconstruction/reference/group/config-params

mcopejans replied:
So it would be better to rename the field name to FLOORPLAN_MAX_ERROR_PARAM_NAME instead?

smits replied:
I don't think so as it is still part of the Tango_3dr functionality

mcopejans commented on tango_ros_common/tango_ros_native/include/tango_ros_native/tango_ros_node.h:
```diff
@@ -81,15 +84,25 @@ const std::string USE_FLOOR_PLAN_PARAM_NAME = "use_floor_plan";
 const std::string ENABLE_DEPTH = "enable_depth";
 const std::string ENABLE_COLOR_CAMERA = "enable_color_camera";
```
Nit. add `_PARAM_NAME` suffix for consistency.

smits replied:
+1

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_conversions_helper.cpp:
```diff
@@ -278,6 +278,8 @@ void toOccupancyGrid(const Tango3DR_ImageBuffer& image_grid,
       int value = static_cast<int>(image_grid.data[j + (image_grid.height - i - 1) * image_grid.width]);
+      occupancy_grid->data.push_back(value);
+      continue;
       if (value == 1) {
```
I would prefer to comment out the code here instead of putting continue to make it very clear that it not used anymore.

smits replied:
+1

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -1013,52 +1083,71 @@ void TangoRosNode::PublishMesh() {
+      get_and_publish_occupancy_grid(occupancy_grid_space_publisher_,
+                                     TANGO_3DR_LAYER_SPACE);
+      get_and_publish_occupancy_grid(occupancy_grid_walls_publisher_,
+                                           TANGO_3DR_LAYER_WALLS);
```
Nit. formatting.

smits replied:
+1

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -1013,52 +1083,71 @@ void TangoRosNode::PublishMesh() {
-          Tango3DR_ImageBuffer image_grid;
-          result = Tango3DR_extractFullFloorplanImage(
-              t3dr_context_, TANGO_3DR_LAYER_OBSTACLES, &origin, &image_grid);
+      auto get_and_publish_occupancy_grid = [&] (ros::Publisher& publisher,
```
Can publisher be const?

smits replied:
+1

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -1013,52 +1083,71 @@ void TangoRosNode::PublishMesh() {
-          result = Tango3DR_extractFullFloorplanImage(
-              t3dr_context_, TANGO_3DR_LAYER_OBSTACLES, &origin, &image_grid);
+      auto get_and_publish_occupancy_grid = [&] (ros::Publisher& publisher,
+          Tango3DR_FloorplanLayer layer) {
```
Const ref?

smits replied:
It's an enum, const ref looks like overkill to me.

mcopejans commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -1013,52 +1083,71 @@ void TangoRosNode::PublishMesh() {
-          LOG(ERROR) << "Tango3DR_updateFullFloorplan failed with error code: " << result;
         }
-      }
+      };
```
Converting above function to a private helper function would make the code easier to read, but this is just my opinion.

smits replied:
As it is not used anywhere else the lambda seems like a nice solution to me.

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_conversions_helper.cpp:
```diff
@@ -278,6 +278,10 @@ void toOccupancyGrid(const Tango3DR_ImageBuffer& image_grid,
       int value = static_cast<int>(image_grid.data[j + (image_grid.height - i - 1) * image_grid.width]);
+      // Push raw data into message.
+      occupancy_grid->data.push_back(value);
+      /* Disabling this conversion for now
```
Is this disabled for testing purposes?
From the [documentation](http://docs.ros.org/kinetic/api/nav_msgs/html/msg/OccupancyGrid.html) an occupancy grid expects value between 0 and 100.

smits replied:
I'm trying to find a generic solution here for all layers, the obstacle layer is playing dirty.

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
   }
+
+  node_handle_.param(TANGO_3DR_USE_SPACE_CLEARING_PARAM_NAME,
+                     t3dr_resolution_, TANGO_3DR_DEFAULT_USE_SPACE_CLEARING);
```
nit: Make  new variable t3dr_use_space_clearing.

smits replied:
this explains so much :blush: , I already wondered why playing with the parameters did not really change anything

smits replied:
done

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+  node_handle_.param(TANGO_3DR_USE_SPACE_CLEARING_PARAM_NAME,
+                     t3dr_resolution_, TANGO_3DR_DEFAULT_USE_SPACE_CLEARING);
+  const char* use_space_clearing = "use_space_clearing";
+  result = Tango3DR_Config_setBool(t3dr_config, use_space_clearing, true);
```
nit: t3dr_use_space_clearing

smits replied:
done

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+    return result;
+  }
+  node_handle_.param(TANGO_3DR_RESOLUTION_PARAM_NAME,
+                     t3dr_resolution_, TANGO_3DR_DEFAULT_RESOLUTION);
```
nit: TANGO_3DR_MIN_NUM_VERTICES + make new variable t3dr_min_num_vertices.

smits replied:
done

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+  node_handle_.param(TANGO_3DR_RESOLUTION_PARAM_NAME,
+                     t3dr_resolution_, TANGO_3DR_DEFAULT_RESOLUTION);
+  const char* min_num_vertices = "min_num_vertices";
+  result = Tango3DR_Config_setInt32(t3dr_config, min_num_vertices, 20);
```
nit: t3dr_min_num_vertices

smits replied:
done

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+    return result;
+  }
+  node_handle_.param(TANGO_3DR_RESOLUTION_PARAM_NAME,
+                     t3dr_resolution_, TANGO_3DR_DEFAULT_RESOLUTION);
```
Same here.

smits replied:
done

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+  node_handle_.param(TANGO_3DR_RESOLUTION_PARAM_NAME,
+                     t3dr_resolution_, TANGO_3DR_DEFAULT_RESOLUTION);
+  const char* update_method = "update_method";
+  result = Tango3DR_Config_setInt32(t3dr_config, update_method, TANGO_3DR_TRAVERSAL_UPDATE);
```
Same here.

smits replied:
done

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+        << update_method << " error: " << result;
+    return result;
+  }
+  node_handle_.param(TANGO_3DR_RESOLUTION_PARAM_NAME,
```
Same here.

smits replied:
done

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+  node_handle_.param(TANGO_3DR_RESOLUTION_PARAM_NAME,
+                     t3dr_resolution_, TANGO_3DR_DEFAULT_RESOLUTION);
+  const char* max_voxel_weight = "max_voxel_weight";
+  result = Tango3DR_Config_setInt32(t3dr_config, max_voxel_weight, 1000);
```
Same here.

smits replied:
done

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+    return result;
+  }
+  node_handle_.param(TANGO_3DR_RESOLUTION_PARAM_NAME,
+                     t3dr_resolution_, TANGO_3DR_DEFAULT_RESOLUTION);
```
Same here.

smits replied:
done

PerrineAguiar commented on tango_ros_common/tango_ros_native/src/tango_ros_node.cpp:
```diff
@@ -446,6 +473,53 @@ Tango3DR_Status TangoRosNode::TangoSetup3DRConfig() {
+  node_handle_.param(TANGO_3DR_RESOLUTION_PARAM_NAME,
+                     t3dr_resolution_, TANGO_3DR_DEFAULT_RESOLUTION);
+  const char* floorplan_max_error = "floorplan_max_error";
+  result = Tango3DR_Config_setDouble(t3dr_config, floorplan_max_error, 0.1);
```
Same here.

smits replied:
done

smits commented:
@PerrineAguiar addressed most of your comments, we just need to come up with a nice generic solution for the occupancy grid generation from the floorplan layers

smits commented:
I propose to close this PR, we'll only expose the obstacle layer as map and just iterate on that one to make it good enough. We should cherrypick some commits in a separate PR to get the additional configuration options. @PerrineAguiar Can you take care of this.
